File: src/eQuantic.UI.Core/HtmlStyle.cs
using System.Collections.Generic;
using System.Linq;

namespace eQuantic.UI.Core;

/// <summary>
/// Inline CSS styles with type-safe properties
/// </summary>
public class HtmlStyle
{
    #region Layout

    public Display? Display { get; set; }
    public Position? Position { get; set; }
    public string? Top { get; set; }
    public string? Right { get; set; }
    public string? Bottom { get; set; }
    public string? Left { get; set; }
    public string? ZIndex { get; set; }

    #endregion

    #region Flexbox

    public FlexDirection? FlexDirection { get; set; }
    public FlexWrap? FlexWrap { get; set; }
    public JustifyContent? JustifyContent { get; set; }
    public AlignItem? AlignItems { get; set; }
    public AlignItem? AlignContent { get; set; }
    public string? Gap { get; set; }
    public string? Flex { get; set; }
    public string? FlexGrow { get; set; }
    public string? FlexShrink { get; set; }
    public string? AlignSelf { get; set; }

    #endregion

    #region Grid

    public string? GridTemplateColumns { get; set; }
    public string? GridTemplateRows { get; set; }
    public string? GridColumn { get; set; }
    public string? GridRow { get; set; }
    public GridFlow? GridAutoFlow { get; set; }
    public JustifyContent? JustifyItems { get; set; }

    #endregion

    #region Sizing

    public string? Width { get; set; }
    public string? Height { get; set; }
    public string? AspectRatio { get; set; }
    public string? MinWidth { get; set; }
    public string? MinHeight { get; set; }
    public string? MaxWidth { get; set; }
    public string? MaxHeight { get; set; }

    #endregion

    #region Spacing

    public string? Margin { get; set; }
    public string? MarginTop { get; set; }
    public string? MarginRight { get; set; }
    public string? MarginBottom { get; set; }
    public string? MarginLeft { get; set; }

    public string? Padding { get; set; }
    public string? PaddingTop { get; set; }
    public string? PaddingRight { get; set; }
    public string? PaddingBottom { get; set; }
    public string? PaddingLeft { get; set; }

    #endregion

    #region Background

    public string? Background { get; set; }
    public string? BackgroundColor { get; set; }
    public string? BackgroundImage { get; set; }
    public string? BackgroundSize { get; set; }
    public string? BackgroundPosition { get; set; }

    /// <summary>`background-clip` — `text` paints the background through the glyphs (gradient text).
    /// The `-webkit-` twin is emitted alongside it: Safari still requires the prefix.</summary>
    public string? BackgroundClip { get; set; }

    /// <summary>`-webkit-text-fill-color` — set to `transparent` so the clipped background shows
    /// through. Prefixed-only by specification; there is no unprefixed equivalent.</summary>
    public string? WebkitTextFillColor { get; set; }

    public string? Filter { get; set; }

    #endregion

    #region Border

    public string? Border { get; set; }
    public string? BorderWidth { get; set; }
    public string? BorderStyle { get; set; }
    public string? BorderColor { get; set; }
    public string? BorderRadius { get; set; }

    #endregion

    public string? ObjectFit { get; set; }

    /// <summary>CSS custom properties (<c>--name: value</c>), emitted AFTER every ordered property —
    /// per-element inputs for generated-stylesheet mechanics (e.g. the pressed-state swap).</summary>
    public Dictionary<string, string>? CustomProperties { get; set; }
    public string? ObjectPosition { get; set; }

    #region Typography

    public string? Color { get; set; }
    public string? FontFamily { get; set; }
    public string? FontSize { get; set; }
    public string? FontWeight { get; set; }
    public string? FontStyle { get; set; }
    public string? LineHeight { get; set; }
    public TextAlign? TextAlign { get; set; }
    public string? TextDecoration { get; set; }
    public string? TextTransform { get; set; }
    public string? LetterSpacing { get; set; }

    #endregion

    #region Effects

    public string? BoxShadow { get; set; }

    /// <summary>Emitted as BOTH <c>backdrop-filter</c> and <c>-webkit-backdrop-filter</c> — Safari
    /// still requires the prefix, and the atomizer dedupes each declaration independently.</summary>
    public string? BackdropFilter { get; set; }

    /// <summary>Multi-line clamp: the axis the legacy flexbox lays out on (`vertical`).</summary>
    public string? WebkitBoxOrient { get; set; }

    /// <summary>Multi-line clamp: how many lines survive before the ellipsis.</summary>
    public string? WebkitLineClamp { get; set; }

    public string? Opacity { get; set; }
    public string? Visibility { get; set; }
    public string? PointerEvents { get; set; }
    public string? Cursor { get; set; }
    public string? Overflow { get; set; }
    public string? OverflowX { get; set; }
    public string? OverflowY { get; set; }
    public string? Transition { get; set; }
    public string? Transform { get; set; }
    public string? Animation { get; set; }
    public string? AnimationDelay { get; set; }
    public string? WhiteSpace { get; set; }
    public string? TextOverflow { get; set; }
    public string? BoxSizing { get; set; }

    public string? PlaceItems { get; set; }

    public string? GridArea { get; set; }

    #endregion

    /// <summary>
    /// The style's REGULAR declarations as ordered (property, value) pairs — the seam the web
    /// realizer's StyleAtomizer consumes (each pair becomes one deduplicated atomic class).
    /// Custom properties are deliberately NOT included: they are per-element inputs (tier 3 of the
    /// style pipeline) and stay inline; read them from <see cref="CustomProperties"/>.
    /// </summary>
    public IReadOnlyList<KeyValuePair<string, string>> EnumerateDeclarations()
    {
        var properties = new List<KeyValuePair<string, string>>();

        AddProperty(properties, "display", Display);
        AddProperty(properties, "position", Position);
        AddProperty(properties, "top", Top);
        AddProperty(properties, "right", Right);
        AddProperty(properties, "bottom", Bottom);
        AddProperty(properties, "left", Left);
        AddProperty(properties, "place-items", PlaceItems);
        AddProperty(properties, "grid-area", GridArea);
        AddProperty(properties, "z-index", ZIndex);

        AddProperty(properties, "flex-direction", FlexDirection);
        AddProperty(properties, "flex-wrap", FlexWrap);
        AddProperty(properties, "justify-content", JustifyContent);
        AddProperty(properties, "align-items", AlignItems);
        AddProperty(properties, "align-content", AlignContent);
        AddProperty(properties, "gap", Gap);
        AddProperty(properties, "flex", Flex);
        AddProperty(properties, "flex-grow", FlexGrow);
        AddProperty(properties, "flex-shrink", FlexShrink);
        AddProperty(properties, "align-self", AlignSelf);

        AddProperty(properties, "grid-template-columns", GridTemplateColumns);
        AddProperty(properties, "grid-template-rows", GridTemplateRows);
        AddProperty(properties, "grid-column", GridColumn);
        AddProperty(properties, "grid-row", GridRow);
        AddProperty(properties, "grid-auto-flow", GridAutoFlow);
        AddProperty(properties, "justify-items", JustifyItems);

        AddProperty(properties, "width", Width);
        AddProperty(properties, "height", Height);
        AddProperty(properties, "min-width", MinWidth);
        AddProperty(properties, "min-height", MinHeight);
        AddProperty(properties, "max-width", MaxWidth);
        AddProperty(properties, "max-height", MaxHeight);
        AddProperty(properties, "aspect-ratio", AspectRatio);

        AddProperty(properties, "margin", Margin);
        AddProperty(properties, "margin-top", MarginTop);
        AddProperty(properties, "margin-right", MarginRight);
        AddProperty(properties, "margin-bottom", MarginBottom);
        AddProperty(properties, "margin-left", MarginLeft);

        AddProperty(properties, "padding", Padding);
        AddProperty(properties, "padding-top", PaddingTop);
        AddProperty(properties, "padding-right", PaddingRight);
        AddProperty(properties, "padding-bottom", PaddingBottom);
        AddProperty(properties, "padding-left", PaddingLeft);

        AddProperty(properties, "background", Background);
        AddProperty(properties, "background-color", BackgroundColor);
        AddProperty(properties, "background-image", BackgroundImage);
        AddProperty(properties, "background-size", BackgroundSize);
        AddProperty(properties, "background-position", BackgroundPosition);
        AddProperty(properties, "-webkit-background-clip", BackgroundClip);
        AddProperty(properties, "background-clip", BackgroundClip);
        AddProperty(properties, "-webkit-text-fill-color", WebkitTextFillColor);
        AddProperty(properties, "filter", Filter);

        AddProperty(properties, "border", Border);
        AddProperty(properties, "border-width", BorderWidth);
        AddProperty(properties, "border-style", BorderStyle);
        AddProperty(properties, "border-color", BorderColor);
        AddProperty(properties, "border-radius", BorderRadius);

        AddProperty(properties, "object-fit", ObjectFit);
        AddProperty(properties, "object-position", ObjectPosition);

        AddProperty(properties, "color", Color);
        AddProperty(properties, "font-family", FontFamily);
        AddProperty(properties, "font-size", FontSize);
        AddProperty(properties, "font-weight", FontWeight);
        AddProperty(properties, "font-style", FontStyle);
        AddProperty(properties, "line-height", LineHeight);
        AddProperty(properties, "text-align", TextAlign);
        AddProperty(properties, "text-decoration", TextDecoration);
        AddProperty(properties, "text-transform", TextTransform);
        AddProperty(properties, "letter-spacing", LetterSpacing);

        AddProperty(properties, "box-shadow", BoxShadow);
        AddProperty(properties, "-webkit-box-orient", WebkitBoxOrient);
        AddProperty(properties, "-webkit-line-clamp", WebkitLineClamp);
        AddProperty(properties, "opacity", Opacity);
        AddProperty(properties, "visibility", Visibility);
        AddProperty(properties, "pointer-events", PointerEvents);
        AddProperty(properties, "backdrop-filter", BackdropFilter);
        AddProperty(properties, "-webkit-backdrop-filter", BackdropFilter);
        AddProperty(properties, "cursor", Cursor);
        AddProperty(properties, "overflow", Overflow);
        AddProperty(properties, "overflow-x", OverflowX);
        AddProperty(properties, "overflow-y", OverflowY);
        AddProperty(properties, "transition", Transition);
        AddProperty(properties, "transform", Transform);
        AddProperty(properties, "animation", Animation);
        AddProperty(properties, "animation-delay", AnimationDelay);
        AddProperty(properties, "white-space", WhiteSpace);
        AddProperty(properties, "text-overflow", TextOverflow);
        AddProperty(properties, "box-sizing", BoxSizing);

        return properties;
    }

    /// <summary>
    /// Convert to CSS string for inline styles
    /// </summary>
    public string ToCssString()
    {
        var properties = EnumerateDeclarations()
            .Select(p => $"{p.Key}: {p.Value}")
            .ToList();

        // Custom properties come LAST — per-element inputs for generated-stylesheet mechanics
        // (e.g. --eq-pressed-bg); the tail position is part of the hydration cross-pin.
        if (CustomProperties != null)
        {
            foreach (var custom in CustomProperties)
                properties.Add($"{custom.Key}: {custom.Value}");
        }

        return string.Join("; ", properties);
    }

    private static void AddProperty(List<KeyValuePair<string, string>> properties, string name, object? value)
    {
        if (value == null) return;

        string cssValue;
        if (value is Enum)
        {
            // Enum member names (e.g. SpaceBetween) map to lowercase, hyphenated CSS keywords.
            cssValue = value.ToString()!.ToLowerInvariant()
                .Replace("flexstart", "flex-start")
                .Replace("flexend", "flex-end")
                .Replace("inlineblock", "inline-block")
                .Replace("inlineflex", "inline-flex")
                .Replace("inlinegrid", "inline-grid")
                .Replace("spacebetween", "space-between")
                .Replace("spacearound", "space-around")
                .Replace("spaceevenly", "space-evenly")
                .Replace("rowreverse", "row-reverse")
                .Replace("columnreverse", "column-reverse")
                .Replace("wrapreverse", "wrap-reverse")
                .Replace("rowdense", "row dense")
                .Replace("columndense", "column dense")
                .Replace("webkitbox", "-webkit-box");
        }
        else
        {
            // Arbitrary values — URLs, data: URIs, base64, custom properties — are
            // case-sensitive and must be preserved verbatim (never lowercased).
            cssValue = value.ToString()!;
        }

        properties.Add(new KeyValuePair<string, string>(name, cssValue));
    }
}

File: src/eQuantic.UI.Runtime/src/core/html-style.ts
/**
 * Inline CSS styles with type-safe properties
 */
export class HtmlStyle {
  // Layout
  display?: string;
  position?: string;
  top?: string;
  right?: string;
  bottom?: string;
  left?: string;
  zIndex?: string;

  // Flexbox
  flexDirection?: string;
  flexWrap?: string;
  justifyContent?: string;
  alignItems?: string;
  alignContent?: string;
  gap?: string;
  flex?: string;
  flexGrow?: string;
  flexShrink?: string;

  // Grid
  gridTemplateColumns?: string;
  gridTemplateRows?: string;
  gridColumn?: string;
  gridRow?: string;
  gridAutoFlow?: string;
  justifyItems?: string;

  // Sizing
  width?: string;
  height?: string;
  minWidth?: string;
  minHeight?: string;
  maxWidth?: string;
  maxHeight?: string;

  // Spacing
  margin?: string;
  marginTop?: string;
  marginRight?: string;
  marginBottom?: string;
  marginLeft?: string;
  padding?: string;
  paddingTop?: string;
  paddingRight?: string;
  paddingBottom?: string;
  paddingLeft?: string;

  // Background
  background?: string;
  backgroundColor?: string;
  backgroundImage?: string;

  // Border
  border?: string;
  borderWidth?: string;
  borderStyle?: string;
  borderColor?: string;
  borderRadius?: string;

  // Typography
  color?: string;
  fontFamily?: string;
  fontSize?: string;
  fontWeight?: string;
  fontStyle?: string;
  lineHeight?: string;
  textAlign?: string;
  textDecoration?: string;
  textTransform?: string;
  letterSpacing?: string;

  // Effects
  boxShadow?: string;
  opacity?: string;
  cursor?: string;
  overflow?: string;
  overflowX?: string;
  overflowY?: string;
  transition?: string;
  transform?: string;

  constructor(init?: Partial<HtmlStyle>) {
    if (init) {
      Object.assign(this, init);
    }
  }

  /**
   * Convert to CSS string for inline styles
   */
  toCssString(): string {
    const properties: string[] = [];
    const cssMap: Record<string, string> = {
      // Convert camelCase to kebab-case
      display: 'display',
      position: 'position',
      top: 'top',
      right: 'right',
      bottom: 'bottom',
      left: 'left',
      zIndex: 'z-index',
      flexDirection: 'flex-direction',
      flexWrap: 'flex-wrap',
      justifyContent: 'justify-content',
      alignItems: 'align-items',
      alignContent: 'align-content',
      gap: 'gap',
      flex: 'flex',
      flexGrow: 'flex-grow',
      flexShrink: 'flex-shrink',
      gridTemplateColumns: 'grid-template-columns',
      gridTemplateRows: 'grid-template-rows',
      gridColumn: 'grid-column',
      gridRow: 'grid-row',
      gridAutoFlow: 'grid-auto-flow',
      justifyItems: 'justify-items',
      width: 'width',
      height: 'height',
      minWidth: 'min-width',
      minHeight: 'min-height',
      maxWidth: 'max-width',
      maxHeight: 'max-height',
      margin: 'margin',
      marginTop: 'margin-top',
      marginRight: 'margin-right',
      marginBottom: 'margin-bottom',
      marginLeft: 'margin-left',
      padding: 'padding',
      paddingTop: 'padding-top',
      paddingRight: 'padding-right',
      paddingBottom: 'padding-bottom',
      paddingLeft: 'padding-left',
      background: 'background',
      backgroundColor: 'background-color',
      backgroundImage: 'background-image',
      border: 'border',
      borderWidth: 'border-width',
      borderStyle: 'border-style',
      borderColor: 'border-color',
      borderRadius: 'border-radius',
      color: 'color',
      fontFamily: 'font-family',
      fontSize: 'font-size',
      fontWeight: 'font-weight',
      fontStyle: 'font-style',
      lineHeight: 'line-height',
      textAlign: 'text-align',
      textDecoration: 'text-decoration',
      textTransform: 'text-transform',
      letterSpacing: 'letter-spacing',
      boxShadow: 'box-shadow',
      opacity: 'opacity',
      cursor: 'cursor',
      overflow: 'overflow',
      overflowX: 'overflow-x',
      overflowY: 'overflow-y',
      transition: 'transition',
      transform: 'transform',
    };

    for (const [key, cssName] of Object.entries(cssMap)) {
      const value = (this as any)[key];
      if (value !== undefined && value !== null) {
        properties.push(`${cssName}: ${value}`);
      }
    }

    return properties.join('; ');
  }
}
